### ISSUE TITLE
Initializing iOS SDK from code

### ISSUE THREAD

**Issue by perholmes** [getsentry/sentry-unity#303]
Hi,

Currently, if you want to use the iOS SDK, you're forced to use the (limited) Sentry control panel in Unity, which precludes the use of countless features, like hooks. The documentation mentions that the documentation is missing for how to initialize the iOS SDK in code: https://docs.sentry.io/platforms/unity/configuration/ios-native/

When is this documentation due? It's kind of a deal-breaker because the choice is between using a severely limited version of Sentry for Unity, *OR* have iOS crash handling support. Neither is tolerable.

**Comment by bruno-garcia** [comment]
We don't have an ETA for a complete docs on how to go manual yet.

**Comment by patrickkulling** [comment]
@bruno-garcia 

I stumbled upon the same "issue" and was wondering about this as well.

From a developer perspective I prefer the code option because it's more intuitive to search for it and search for changes than a serialised yaml file.

Does anything speak against always adding the Sentry.framework and so on (essentially what the PostProcessor is doing), but configure the rest via code ?

```
SentryUnity.Init (c =>
      {
        c.AndroidNativeSupportEnabled = true;
        c.IosNativeSupportEnabled = true;
        c.Dsn = "xxx";
        c.DiagnosticLevel = SentryLevel.Error;
        c.MaxBreadcrumbs = 150;
        c.AttachStacktrace = true;
        c.Release = ClientConfig.Instance.Version.ToString ();
      });
```

## A real use case is the following

Although we have one Unity codebase for all the platforms that we support (iOS, Google Play, Amazon, WebGL) we want to be able to use different dsn's per platform instead of using a tag.

The main reason for that is that we get a better picture of error trends, releases, can configure different inbound filters, ...

I could see an additional use case where one would also want to use a different sentry project not only per platform but per environment (e.g. ci, staging, production) and use different filters and sampling rates. 

The current configuration via the Editor window is just too limiting.

Best,
Patrick

**Comment by bitsandfoxes** [comment]
Hey @patrickkulling, we see the use case and while we do intend to enable native support with a programmatic approach we're just not quite there yet.

A bit more context: Right now the PostProcessor does a couple of additional things like 
- Adding Sentry Init to the main.mm
- Writing the options to NativeOptions.m
We do this so the native SDK gets initialized even before Unity itself starts to "catch 'em all". That's why we currently rely on the options config being already available during build time. Any programmatic initialization from the C# layer would automatically happen later.

To unblock you there might be a workaround: We recently made all the properties on the ScriptableOptions object public. So you could load the ScriptableOptions from the static `GetConfigPath` and overwrite relevant options. Your editor script should run before any build-related post-processing.

**Comment by patrickkulling** [comment]
@bitsandfoxes I have seen this commit and was wondering about this intend.
I think this is fair enough. I also totally understand why the post processor behaves that way it does at the moment. I simply wanted to give more context on this issue since the OP never got back to this.

For my specific use cases this would work good enough!

Best
Patrick

**Comment by perholmes** [comment]
OP here, it wasn't just about hooks. I'm not the middle of it right now, but I could identify a handful of parameters that weren't available from code.

I'm not clear in which way the ScriptableSentryUnityOptions are different from SentryUnityOptions. Are they to be used instead of? Many of the parameters are the same. Or are the scriptable options defaults that the runtime Sentry.Init options override? It's not clear what relationship and priority these parameters have.

What has to be provided in ScriptableSentryUnityOptions at build-time? I assume enabling iOS native support and a DSN at a minimum? And how are these provided to the SDK at build-time? Is there a static function to call?

But either way, an editor script is a non-starter for us. Manual steps lead to botched releases (speaking from rich experience). In other projects, manual build steps have led to catastrophic releases, and they greatly discourage any kind of ongoing releasing, because a release becomes like a rocket launch that has to be done perfectly or you're dead.

Without code-access to iOS releasing, we'd have to (a) not have native crash handling at all, or (b) ditch Sentry across all our projects on all platforms. It sounds hyperbolic, but this editor script will lead to a catastrophic failure, and it's not an acceptable risk.

**Comment by bitsandfoxes** [comment]
> I'm not clear in which way the ScriptableSentryUnityOptions are different from SentryUnityOptions. Are they to be used instead of? Many of the parameters are the same. Or are the scriptable options defaults that the runtime Sentry.Init options override? It's not clear what relationship and priority these parameters have.

The `ScriptableSentryOptions` is the config file that lives in the Resource folder. The SDK reads the config from there so it can properly initialize the native SDKs during build time. Basically: 

- ScriptableSentryOptions -> Unity Editor & Buildtime config
- SentryOptions -> Runtime (i.e. you initialize the SDK programmatically)

The ScriptableSentryOptions get created when you first go to `Tool->Sentry` and you can tweak the values via the config window. Alternatively, the scriptable object's properties are now public so you can have your own platform-specific scripts access and overwrite it.

> But either way, an editor script is a non-starter for us. Manual steps lead to botched releases (speaking from rich experience). In other projects, manual build steps have led to catastrophic releases, and they greatly discourage any kind of ongoing releasing, because a release becomes like a rocket launch that has to be done perfectly or you're dead.

With editor script, we don't mean something you'd have to click. We mean i.e. [OnPreProcessBuild](https://docs.unity3d.com/ScriptReference/Build.IPreprocessBuildWithReport.OnPreprocessBuild.html) to modify the scriptable object for setting your platform-specific configs.
Can you tell us what steps you currently would have to do manually? So we can take a look?

**Comment by vaind** [comment]
@bitsandfoxes you're probably most aware of this in relation to the [programmatic options initialization in v0.10.1](https://github.com/getsentry/sentry-unity/releases/tag/0.10.1) - can you please give a quick update on this issue's scope and whether something is still missing?

**Comment by bitsandfoxes** [comment]
Initializing the native SDKs from code is still not yet supported but definitely a feature that we will add in the future. What we added with 0.10.1 is support for programmatic configuration. 
So the Unity SDK initializes itself and calls into `Configure` of a provided `ScriptableOptionConfiguration`object. You can have multiple instances of these and swap them out in the configuration window -> Options Config.
The current limitation this has is that your changes to the options object will not be reflected on the native layer as these will be set during build time.

**Comment by vaind** [comment]
I'm looking into this - the approach I'm taking is the same as for [c# scriptable options ](https://github.com/getsentry/sentry-unity/pull/564/files)- allow users to define a script (in the Assets folder), objective-c in case of iOS, that is then loaded and used on top of the default options.

**Comment by vaind** [comment]
As has been said above, configuring options for dotnet is already possible from code. Doing the same for iOS and other native platforms is planned and tracked in another issue. Closing this one in favor of #609
